File: src/test/java/com/droiddevgeeks/crashsink/AndroidDeviceMetadataTest.java
package com.droiddevgeeks.crashsink;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

import org.junit.Test;

public class AndroidDeviceMetadataTest {

    @Test public void collectsAppVersionFromPackageInfo() throws Exception {
        Context context = mock(Context.class);
        PackageManager pm = mock(PackageManager.class);
        PackageInfo pi = new PackageInfo();
        pi.versionName = "1.2.3";
        pi.versionCode = 42;

        when(context.getPackageManager()).thenReturn(pm);
        when(context.getPackageName()).thenReturn("com.x");
        when(pm.getPackageInfo("com.x", 0)).thenReturn(pi);

        DeviceMetadata md = AndroidDeviceMetadata.collect(context);

        assertEquals("1.2.3", md.appVersionName);
        assertEquals(42, md.appVersionCode);
        // Build.* fields are default/null under unit tests; do not assert on them.
    }

    @Test public void leavesDefaultsWhenPackageManagerThrows() throws Exception {
        Context context = mock(Context.class);
        PackageManager pm = mock(PackageManager.class);

        when(context.getPackageManager()).thenReturn(pm);
        when(context.getPackageName()).thenReturn("com.x");
        when(pm.getPackageInfo("com.x", 0))
                .thenThrow(new PackageManager.NameNotFoundException());

        DeviceMetadata md = AndroidDeviceMetadata.collect(context);

        assertEquals("unknown", md.appVersionName);
        assertEquals(0, md.appVersionCode);
    }
}

File: crashsink/src/test/kotlin/com/droiddevgeeks/crashsink/AndroidDeviceMetadataTest.kt
package com.droiddevgeeks.crashsink

import android.content.Context
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import org.junit.Assert.assertEquals
import org.junit.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`

class AndroidDeviceMetadataTest {

    @Test @Suppress("DEPRECATION") fun collectsAppVersionFromPackageInfo() {
        val context = mock(Context::class.java)
        val pm = mock(PackageManager::class.java)
        val pi = PackageInfo()
        pi.versionName = "1.2.3"
        pi.versionCode = 42

        `when`(context.packageManager).thenReturn(pm)
        `when`(context.packageName).thenReturn("com.x")
        `when`(pm.getPackageInfo("com.x", 0)).thenReturn(pi)

        val md = AndroidDeviceMetadata.collect(context)

        assertEquals("1.2.3", md.appVersionName)
        assertEquals(42, md.appVersionCode)
        // Build.* fields are default/null under unit tests; do not assert on them.
    }

    @Test fun leavesDefaultsWhenPackageManagerThrows() {
        val context = mock(Context::class.java)
        val pm = mock(PackageManager::class.java)

        `when`(context.packageManager).thenReturn(pm)
        `when`(context.packageName).thenReturn("com.x")
        `when`(pm.getPackageInfo("com.x", 0)).thenThrow(PackageManager.NameNotFoundException())

        val md = AndroidDeviceMetadata.collect(context)

        assertEquals("unknown", md.appVersionName)
        assertEquals(0, md.appVersionCode)
    }
}
